torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -415,14 +434,15 @@ class ChunkDataset final
   // batches and caches them in batch_buffer_.
   ChunkReader chunk_reader_;
 
-  // chunk sampler to shuffle different chunks
-  samplers::LockedSampler<ChunkSamplerType> chunk_sampler_;
-
   // example sampler to shuffle examples in a specific chunk
   ExampleSamplerType example_sampler_;
 
+  // Selects chunks and their order for this reader.
+  std::shared_ptr<samplers::DistributedSampler> chunk_sampler_;
```

Comment:
uuuuuuh that's a very weird specialization. It's not like `ChunkDataset` is specific to distributed workloads

Reply:
This is the reason for introducing ChunkSelectors originally in the PR. Either, the sampler or ChunkDataset needs to know this is operating under distributed setting. We cannot have the sampler base type here because it does not have all the methods we need. This leads us back to local_chunk_count() method. Let's discuss offline and see if we can resolve this in a better way.